src/Features/CSharp/Portable/Completion/KeywordRecommenders/WhenKeywordRecommender.cs

```diff
@@ -15,26 +18,100 @@ public WhenKeywordRecommender()
 
         protected override bool IsValidContext(int position, CSharpSyntaxContext context, CancellationToken cancellationToken)
         {
-            if (context.IsCatchFilterContext)
+            return context.IsCatchFilterContext ||
+                (IsAfterCompleteExpressionOrPatternInCaseLabel(context, out var expressionOrPattern) &&
+                !IsTypeName(expressionOrPattern, context.SemanticModel, cancellationToken));
+        }
+
+        private static bool IsAfterCompleteExpressionOrPatternInCaseLabel(CSharpSyntaxContext context,
+            out SyntaxNode expressionOrPattern)
+        {
+            var switchLabel = context.TargetToken.GetAncestor<SwitchLabelSyntax>();
+            if (switchLabel == null)
             {
-                return true;
+                expressionOrPattern = null;
+                return false;
             }
 
-            // case int i $$
-            // case int i w$$
-            if (IsIdentifierInCasePattern(context.TargetToken.Parent))
+            expressionOrPattern = switchLabel.ChildNodes().FirstOrDefault();
+            if (expressionOrPattern == null)
             {
-                return true;
+                // It must have been a default label.
+                return false;
             }
 
-            return false;
+            // If the last token is missing, the expression is incomplete - possibly because of missing parentheses,
+            // but not necessarily. We don't want to offer 'when' in those cases. Here are some examples that illustrate this:
+            // case |
+            // case x.|
+            // case 1 + |
+            // case (1 + 1 |
+
+            // Also note that if there's a missing token inside the expression, that's fine and we do offer 'when':
+            // case (1 + ) |
+
+            var lastToken = expressionOrPattern.GetLastToken(includeZeroWidth: true);
+            if (lastToken.IsMissing)
+            {
+                return false;
+            }
+
+            if (lastToken == context.LeftToken && context.LeftToken != context.TargetToken &&
+                expressionOrPattern is DeclarationPatternSyntax declarationPattern)
+            {
+                // case constant w|
```

Comment:
It may be good to have a command handler test for this scenario. If we do offer a completion for `when`, it should be soft, so that it would not annoy a user typing a new name.

See `CSharpCompletionCommandHandlerTests.vb`

Reply:
what about inside `SuggestionModeCompletionProviderTests.cs`? that seems like the right place for this

Reply:
or just specifically for 'when'?

Reply:
`SuggestionModeCompletionProviderTests.cs` would be good, but I'd still recommend some command handler tests, as those are the closest tests to represent end-to-end experience of typing.

Reply:
ok

Reply:
> and Color/Color scenarios
but the existing unit tests already cover all these cases that the keyword isn't recommended when it shouldn't be... why not just add tests that check for suggestion mode when it *is* be offered?

Reply:
For Color/Color scenario, there is one scenario where `when` is offered (`Color` means a constant) and we should offer `when` and also allow typing a new variable name (then `Color` changes meaning to refer to the type `Color`).
To get good confidence on scenarios where multiple features interact (recommenders and suggestion mode), I'd do a command handler test.

Reply:
I added some tests, could you take a look if you think there's other cases I should add as well?

Reply:
@jcouv tests added. Could you take a look if that's sufficient or if you have some other cases that you'd like me to add as well? Thanks